src/System.Management.Automation/engine/Utils.cs

```diff
@@ -478,13 +478,22 @@ internal static string GetWindowsPowerShellVersionFromRegistry()
         }
 #endif
 
-        internal static string DefaultPowerShellAppBase => GetApplicationBase(DefaultPowerShellShellID);
+        internal static string DefaultPowerShellAppBase => GetApplicationBase();
 
-        internal static string GetApplicationBase(string shellId)
+        internal static string GetApplicationBase()
         {
-            // Use the location of SMA.dll as the application base.
-            Assembly assembly = typeof(PSObject).Assembly;
-            return Path.GetDirectoryName(assembly.Location);
+            // AppContext is needed where PS7 may be compiled as a single exe
+            // Environment.ProcessPath covers the general case
+            if (!string.IsNullOrEmpty(AppContext.BaseDirectory))
+            {
+                return Path.TrimEndingDirectorySeparator(AppContext.BaseDirectory);
+            }
+            else if (!string.IsNullOrEmpty(Path.GetDirectoryName(typeof(PSObject).Assembly.Location)))
+            {
+                return Path.GetDirectoryName(typeof(PSObject).Assembly.Location);
+            }
```

Comment:
We should use a static variable to hold the resulted application base path, instead of running this every time.
Also, users can change the application path by `AppContext.SetData("APP_CONTEXT_BASE_DIRECTORY", <new-path>)`, which may completely break PowerShell, so we have to cache the original value on the first call to `GetApplicationBase`.